hazelcast/include/hazelcast/client/sql/impl/read_optimized_lru_cache.h

```diff
@@ -0,0 +1,233 @@
+        }
+        if (cleanup_threshold <= capacity) {
+            BOOST_THROW_EXCEPTION(client::exception::illegal_argument(
+              "cleanupThreshold <= capacity"));
+        }
+
+        capacity_ = capacity;
+        cleanup_threshold_ = cleanup_threshold;
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * @param default_value the default value if the key is not cached.
+     * @returns Returns the value to which the specified key is cached,
+     * or default value if this cache contains no mapping for the key.
+     */
+    std::shared_ptr<V> get_or_default(const K& key,
+                                      const std::shared_ptr<V>& default_value)
+    {
+        const auto existing_value = get(key);
+        return (existing_value != nullptr) ? existing_value : default_value;
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * Returns the value to which the specified key is cached,
+     * or {@code null} if this cache contains no mapping for the key.
+     * @returns Returns the value to which the specified key is cached
+     */
+    std::shared_ptr<V> get(const K& key)
+    {
+        auto value_from_cache = cache_.get(key);
+        if (value_from_cache == nullptr) {
+            return nullptr;
+        }
+        value_from_cache->touch();
+        return std::make_shared<int32_t>(value_from_cache->value_);
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * @param value the value of the cache entry
+     * @throws exception::illegal_argument if the value equals to nullptr
+     */
+    void put(const K& key, const std::shared_ptr<V>& value)
+    {
+        if (value == nullptr) {
+            BOOST_THROW_EXCEPTION(client::exception::illegal_argument(
+              "Null values are disallowed"));
+        }
+
+        auto old_value =
+          cache_.put(key, std::make_shared<value_and_timestamp<V>>(*value));
+        if (old_value == nullptr && cache_.size() > cleanup_threshold_) {
+            do_cleanup();
+        }
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * Removes the cached value for the given key
+     */
+    void remove(const K& key) { cache_.remove(key); }
+
+protected:
+    /**
+     * Helper class to for simulation atomic lock with RAII
+     */
+    class custom_atomic_lock
+    {
+    public:
+        explicit custom_atomic_lock(bool initial_value = false)
+          : lock_(initial_value)
+        {
+        }
+
+        custom_atomic_lock(const custom_atomic_lock&) = delete;
+        const custom_atomic_lock& operator=(const custom_atomic_lock&) = delete;
+
+        bool try_lock()
+        {
+            bool expected = false;
+            return lock_.compare_exchange_strong(expected, true);
+        }
+
+        void release() { lock_.store(false); }
+
+        ~custom_atomic_lock() { release(); }
+
+    private:
+        std::atomic<bool> lock_;
+    };
```

Comment:
I don't think this class is a must. Raw `std::atomic<bool>` can be used instead.

Reply:
Fixed